boot/seal.go

```diff
@@ -25,65 +25,221 @@ import (
+	if !ok {
+		return nil, fmt.Errorf("bootloader doesn't support trusted assets")
+	}
+
+	for _, system := range systems {
+		// get the command line
+		cmdline, err := ComposeRecoveryCommandLine(model, system)
+		if err != nil {
+			return nil, fmt.Errorf("cannot obtain recovery kernel command line: %v", err)
+		}
+
+		// get kernel information from seed
+		perf := timings.New(nil)
+		_, snaps, err := seedReadSystemEssential(dirs.SnapSeedDir, system, []snap.Type{snap.TypeKernel}, perf)
+		if err != nil {
+			return nil, err
+		}
+		if len(snaps) != 1 {
+			return nil, fmt.Errorf("cannot obtain recovery kernel snap")
+		}
+		seedKernel := snaps[0]
+
+		var kernelRev string
+		if seedKernel.SideInfo.Revision.Store() {
+			kernelRev = seedKernel.SideInfo.Revision.String()
+		}
+
+		recoveryBootChain, err := tbl.RecoveryBootChain(seedKernel.Path)
+		if err != nil {
+			return nil, err
+		}
+
+		// get asset chains
+		assetChain, kbf, err := buildBootAssets(recoveryBootChain, modeenv)
+		if err != nil {
+			return nil, err
+		}
+
+		chains = append(chains, bootChain{
+			BrandID:        model.BrandID(),
+			Model:          model.Model(),
+			Grade:          model.Grade(),
+			ModelSignKeyID: model.SignKeyID(),
+			AssetChain:     assetChain,
+			Kernel:         seedKernel.SnapName(),
+			KernelRevision: kernelRev,
+			KernelCmdlines: []string{cmdline},
+			model:          model,
+			kernelBootFile: kbf,
+		})
+	}
+	return chains, nil
+}
+
+func buildRunModeBootChains(rbl, bl bootloader.Bootloader, model *asserts.Model, modeenv *Modeenv, cmdline string) ([]bootChain, error) {
+	tbl, ok := rbl.(bootloader.TrustedAssetsBootloader)
+	if !ok {
+		return nil, fmt.Errorf("recovery bootloader doesn't support trusted assets")
+	}
+
+	chains := make([]bootChain, 0, len(modeenv.CurrentKernels))
+	for _, k := range modeenv.CurrentKernels {
+		info, err := snap.ParsePlaceInfoFromSnapFileName(k)
+		if err != nil {
+			return nil, err
+		}
+		runModeBootChain, err := tbl.BootChain(bl, info.MountFile())
+		if err != nil {
+			return nil, err
+		}
+
+		// get asset chains
+		assetChain, kbf, err := buildBootAssets(runModeBootChain, modeenv)
+		if err != nil {
+			return nil, err
+		}
+		var kernelRev string
+		if info.SnapRevision().Store() {
+			kernelRev = info.SnapRevision().String()
+		}
+		chains = append(chains, bootChain{
+			BrandID:        model.BrandID(),
+			Model:          model.Model(),
+			Grade:          model.Grade(),
+			ModelSignKeyID: model.SignKeyID(),
+			AssetChain:     assetChain,
+			Kernel:         info.SnapName(),
+			KernelRevision: kernelRev,
+			KernelCmdlines: []string{cmdline},
+			model:          model,
+			kernelBootFile: kbf,
+		})
+	}
+	return chains, nil
+}
+
+func buildBootAssets(bootFiles []bootloader.BootFile, modeenv *Modeenv) (assets []bootAsset, kernel bootloader.BootFile, err error) {
+	assets = make([]bootAsset, len(bootFiles)-1)
+
+	for i, bf := range bootFiles[:len(bootFiles)-1] {
```

Comment:
why do we not include the last bootFile from `bootFiles` here? I would have expected this to be

```suggestion
	for i, bf := range bootFiles {
```

Reply:
should probably have a comment if this was intentional

Reply:
it's intentional, the last bootFile corresponds to the kernel, is not a bootloader asset in itself.

If you read the function you can see it's returned separately in return argument called kernelBootFile. Arguably a different approach would have that splitting out done in the caller, so that this does operate on all its input

Reply:
the easiest might to add a proper doc  comment to this function, it's quite central

Reply:
doc-comment would make the most sense to me